elasticdl/python/worker/task_data_service.py

```diff
@@ -27,44 +32,67 @@ def __init__(
             self.data_reader = create_data_reader(data_origin=None)
         self._warm_up_task = None
         self._has_warmed_up = False
+        self._record_failure_tolerance = record_failure_tolerance
 
     def _reset(self):
         """
         Reset pending tasks and record counts
         """
-        self._record_count = 0
+        self._cur_record_count = 0
         self._reported_record_count = 0
-        self._pending_tasks_with_counts = []
+        self._failed_record_count = 0
+        self._pending_tasks = deque()
         self._current_task = None
 
     def get_current_task(self):
         return self._current_task
 
+    def report_task_done(self, task_id, err_msg):
+        self._worker.report_task_result(task_id, err_msg)
+
     def report_record_done(self, count, err_msg=""):
         """
         Report the number of records in the latest processed batch,
         so TaskDataService knows if some pending tasks are finished
         and report_task_result to the master.
-        self._pending_tasks_with_counts[0][0] is the first pending task,
-        self._pending_tasks_with_counts[0][1] is the number of records
-        in this task.
         """
         self._reported_record_count += count
-        if (
-            len(self._pending_tasks_with_counts)
-            and self._reported_record_count
-            >= self._pending_tasks_with_counts[0][1]
-        ):
+        if err_msg:
+            self._failed_record_count += count
+            logger.warning(
+                "Process records of task contains " "error: %s", err_msg
```

Comment:
The error msg is "Process records of task contains " "error: %s"

Maybe unnecessary " " inside?

Reply:
Fixed